### PR TITLE
Add tagging functionality to the array context

### PR DESCRIPTION
**Pull request by inducer**
This is intended to make metadata capture available generically, even if it boils down to a no-op for the PyOpenCL array context. @kaushikcfd, could you review? I'm particularly curious what you think of `tag_axis` and the `CommonSubexpression` tag.

cc @MTCam @nchristensen @alexfikl

### CODE REVIEW

File: meshmode/array_context.py [hunk 1]
```diff
@@ -175,11 +180,22 @@ def __init__(self, array_context):
         self._array_context = array_context
 
 
-class ArrayContext:
+class CommonSubexpression(Tag):
```
Comment (alexfikl) [inline meshmode/array_context.py:183]:
```suggestion
class CommonSubexpressionTag(Tag):
```
? Mostly to differentiate it a bit from the same in `pymbolic`, but I don't feel strongly about it.

File: meshmode/array_context.py [hunk 2]
```diff
@@ -213,9 +229,11 @@ def __init__(self):
     def _get_fake_numpy_namespace(self):
         return _BaseFakeNumpyNamespace(self)
 
+    @abstractmethod
     def empty(self, shape, dtype):
         raise NotImplementedError
```
Comment (alexfikl) [inline meshmode/array_context.py:234]:
```suggestion
        pass
```
? I think abstract methods have their own exception when not implemented?

**Review by alexfikl — COMMENTED** [review]
I'm still very much in the non-lazy land, so don't have much else to say here. But it looks good!

File: meshmode/array_context.py (hunk 2 above)
Comment (alexfikl) [reply meshmode/array_context.py:234]: Ah, you just can't instantiate them at all, so I guess this makes no difference.

**Review by kaushikcfd — APPROVED** [review]
LGTM.

File: meshmode/array_context.py (hunk 1 above)
Comment (inducer) [reply meshmode/array_context.py]: 51d2f20

**Comment by inducer** [comment]
Thanks to the both of you for taking a look!

**Comment by inducer** [comment]
If you're wondering about the CI failures: https://github.com/illinois-ceesd/mirgecom/issues/302

**Pull request merged**